### ISSUE TITLE
PLC0208 should only recommend fixes when iterating over literals

### ISSUE THREAD

**Issue by Skylion007** [astral-sh/ruff#11694]
I tried PLC0208 on my codebase, and found several spots where iterating over a set was clearly intentional. Here is a salient example:
```python
for rg in {dtype.is_floating_point, False}:
```
This will would get corrected to a sequence, when I want this test case to run over `True` and `False` is `dtype.is_floating_point` is true. Otherwise, only False.

Another useful idiom is:
```python
for index in {0, sparse.shape[dim] // 2, sparse.shape[dim] - 1}:
```
Where I want to loop over all unique values. If the `sparse.shape[dim]` is 2, then I only go over index (0, 1) instead of (0, 1, 1). It doesn't seem like this fix can be safe if it doesn't know what it's referencing. At the very least, the fix should be marked unsafe.

**Comment by Skylion007** [comment]
Ugh, looks like https://github.com/astral-sh/ruff/pull/4907 it suppose to cover this, but fails with parameters or getter calls.

**Comment by charliermarsh** [comment]
I think this is different than #4907? My read is that change was about comprehensions and `set` calls. Pylint does flag the cases you've included here, though we could consider changing our rule.

**Comment by rdaysky** [comment]
I’ve also just encountered PLC0208 on code like this:
```
dt: datetime.datetime = ...
slightly_earlier = dt - datetime.timedelta(hours=4)

for day in {dt.date(), slightly_earlier.date()}:
    ...  # could be the same date or two consecutive dates
```
where the intention is for the loop body to handle each distinct date exactly once.

Please consider restricting the rule to literals only.
